### PR TITLE
Add bundle validation.

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
## Description
Add proper validation for bundle requests.
This would narrow down bundle requests we accepting, since we currently don't enforce bundle rules.
http://hl7.org/fhir/bundle.html see `+ Rule:` 

## Related issues
Addresses [AB#81350](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/81350)

## Testing
Unit tests added.

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with Azure API for FHIR if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [x] Tag the PR with Azure Healthcare APIs if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Breaking

### CODE REVIEW

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/BatchTests.cs [hunk 1]
```diff
@@ -108,7 +108,18 @@ public async Task GivenANonBundleResource_WhenSubmittingABatch_ThenBadRequestIsR
         public async Task GivenBundleTypeIsMissing_WhenSubmittingABundle_ThenMethodNotAllowedExceptionIsReturned()
```
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/BatchTests.cs:108]: Nit: The test name should be updated to match what it is now testing.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Validation/ResourceValidator.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            var result = _validator.Validate(resource);
+            var outcomeIssues = result.Issue.OrderBy(x => x.Severity)
+               .Select(issue =>
+                   new OperationOutcomeIssue(
+                       issue.Severity?.ToString(),
+                       issue.Code.ToString(),
+                       diagnostics: issue.Diagnostics,
+                       detailsText: issue.Details?.Text,
+                       detailsCodes: issue.Details?.Coding != null ? new CodableConceptInfo(issue.Details.Coding.Select(x => new Coding(x.System, x.Code, x.Display))) : null,
+                       location: issue.Location.ToArray()))
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Validation/ResourceValidator.cs:44]: Is location a guarantied field or should it also using the inline null check?

**Comment by LTA-Thinking** [comment]
All the changes to the test resource files were to make them pass the new bundle validator, correct?

**Comment by LTA-Thinking** [comment]
I think this should be a major change in semver since it is technically a breaking change. It doesn't feel like one, but we are producing errors where they previously wouldn't be ones.

**Comment by CaitlinV39** [comment]
@Ivanidzo4ka Does this address #1677

**Comment by Ivanidzo4ka** [comment]
Yep

---
In reply to: [921105456](https://github.com/microsoft/fhir-server/pull/2209#issuecomment-921105456) [](http://example.com/codeflow?ancestors=921105456,921105456)

**Comment by brendankowitz** [comment]
Closing for now

**Pull request closed without merging**